src/app/librarian/librarian.component.ts

```diff
@@ -1,15 +1,203 @@
+              });
+            });
+          });
+        },
+        (error) => {
+          debugger;
+        }
+      );
+  }
+
+  loadAllBooks() {
+    this.lmsService
+      .get(`${environment.adminBackendUrl}${environment.readBookUri}`)
+      .subscribe(
+        (res) => {
+          this.totalBooks = res;
+
+        },
+        (error) => {
+          debugger;
+        }
+      );
+  }
+
+  initializeFormGroup() {
+    this.updateBranchForm = new FormGroup({
+      branchName: new FormControl(this.branchName, [
+        Validators.required,
+        Validators.maxLength(45),
+        Validators.minLength(3),
+      ]),
+      branchAddress: new FormControl(this.branchAddress, [
+        Validators.required,
+        Validators.maxLength(45),
+        Validators.minLength(3),
+      ]),
+      bookId: new FormControl(this.bookId),
+      noOfCopies: new FormControl(this.noOfCopies),
+      branchId: new FormControl(this.branchId),
+      books: new FormControl(this.books),
+    });
   }
 
+
+  updateBranch() {
+    const branch = {
+      branchId: this.updateBranchForm.value.branchId,
+      branchName: this.updateBranchForm.value.branchName,
+      branchAddress: this.updateBranchForm.value.branchAddress,
+    }
+    this.lmsService
+      .post(`${environment.libUrl}${environment.updateBanchUri}`, branch)
+      .subscribe(
+        (res) => {
+          console.log(res);
+        },
+        (error) => {
+          console.log(error);
+        }
+      );
+  }
+
+
+  openUpdate(content, obj) {
+    if (obj !== null) {
+      //this is edit/update mode
+      this.updateBranchForm = this.fb.group({
+        branchId: obj.branchId,
+        branchName: obj.branchName,
+        branchAddress: obj.branchAddress,
+      });
+    }
+
+    this.modalRef = this.modalService.open(content);
+    this.modalRef.result.then(
+      (result) => {
+        this.errMsg = "";
+        this.closeResult = `Closed with ${result}`;
+      },
+      (reason) => {
+        this.errMsg = "";
+        this.closeResult = `Dismissed`;
+      }
+    );
+  }
+
+  openAdd(content, obj) {
+    if (obj !== null) {
+      this.addBookCopiesForm = this.fb.group({
+        books: [obj.books],
+        // bookId: [obj.books.bookId],
+        bookId: obj.bookCopies.bookId,
+        branchId: obj.branchId,
+        noOfCopies: obj.bookCopies.noOfCopies,
+        bookCopies: obj.bookCopies,
+      })
+    }
+
+    this.modalRef = this.modalService.open(content);
+    this.modalRef.result.then(
+      (result) => {
+        this.errMsg = "";
+        this.closeResult = `Closed with ${result}`;
+      },
+      (reason) => {
+        this.errMsg = "";
+        this.closeResult = `Dismissed`;
+      }
+    );
+  }
+
+  updateBookCopies() {
+    const bookCopies = {
+      books: this.addBookCopiesForm.value.books,
+      bookId: this.addBookCopiesForm.value.bookId,
+      branchId: this.addBookCopiesForm.value.bookCopies.branchId,
```

Comment:
Is the update copies feature working correctly? It looks like the bookId and branchId will become undefined here, and the bookId will be set to the correct value in line 189, but the branchId will remain undefined. Line 186 should probably be `branchId: this.addBookCopiesForm.value.branchId,` instead.